scio-smb/src/main/java/org/apache/beam/sdk/extensions/smb/SortedBucketSource.java

```diff
@@ -426,49 +436,57 @@ public boolean advance() throws IOException {
           if (keyComparator.compare(entry, minKeyEntry) == 0) {
             final TupleTag tupleTag = entry.getKey();
             int index = resultSchema.getIndex(tupleTag);
-            final List<Object> values = (List<Object>) valueMap.get(index);
-            final Predicate<Object> predicate = predicates[index];
 
             // Track the canonical # buckets of each source that the key is found in.
             // If we find it in a source with a # buckets >= the parallelism of the job,
             // we know that it doesn't need to be re-hashed as it's already in the right bucket.
-            if (acceptKeyGroup == 1) {
-              entry
-                  .getValue()
-                  .getValue()
-                  .forEachRemaining(
-                      v -> {
-                        if (predicate.apply(values, v)) {
-                          values.add(v);
-                        }
-                      });
-            } else if (acceptKeyGroup == -1
-                && (bucketsPerSource.get(tupleTag) >= parallelism
-                    || keyGroupFilter.apply(minKeyEntry.getValue().getKey()))) {
-              entry
-                  .getValue()
-                  .getValue()
-                  .forEachRemaining(
-                      v -> {
-                        if (predicate.apply(values, v)) {
-                          values.add(v);
-                        }
-                      });
+            final boolean emitKeyGroup =
+                acceptKeyGroup == 1
+                    || (acceptKeyGroup != 0 // make sure it hasn't already been ruled out
+                        && (bucketsPerSource.get(tupleTag) >= parallelism
+                            || keyGroupFilter.apply(minKeyEntry.getValue().getKey())));
+
+            // If the user supplies a predicate, we have to materialize the iterable to apply it
+            boolean materialize = (materializeKeyGroup || predicates[index] != null);
+
+            final Predicate<Object> predicate =
+                predicates[index] != null ? predicates[index] : (xs, x) -> true;
+
+            final Iterator<Object> keyGroupIterator =
+                (Iterator<Object>) entry.getValue().getValue();
+
+            if (emitKeyGroup && !materialize) {
+              valueMap.set(
+                  index,
+                  () ->
```

Comment:
What happens if the user re-iterates this `Iterable`? We want it to fail fast instead of emitting nothing right?

Reply:
Also, what if user code doesn't exhaust the iterator? IIRC I tried adding something like that (the "lazy iterable" attempt) but couldn't get it to work.